### DISCUSSION TITLE
userId parameter in killsnoop()

### DISCUSSION THREAD

**Discussion by DineshI-MS** [bpftrace/bpftrace#3176] [General]
Are there any parameters in the killsnoop() which shows the user Id that initiated the kill command? I am looking to differentiate the kill commands initiated by system vs user. If it is not by system, wanted to understand which user initiated the kill command.

**Comment by viktormalik** [comment]
bpftrace scripts do not provide any configuration with CLI arguments. If you need that, have a look at BCC or libbpf tools which usually provide more functionality.

OTOH, the simplicity of the bpftrace language allows you to easily extend the tools by yourself. For instance, adding UID to `killsnoop.bt` is quite simple:
```diff
diff --git a/tools/killsnoop.bt b/tools/killsnoop.bt
index 1be081d2..921f9594 100755
--- a/tools/killsnoop.bt
+++ b/tools/killsnoop.bt
@@ -18,8 +18,8 @@
 BEGIN
 {
        printf("Tracing kill() signals... Hit Ctrl-C to end.\n");
-       printf("%-15s %7s %-16s %4s %6s %s\n",
-               "TIME", "PID", "COMM", "SIG", "TPID", "RESULT");
+       printf("%-15s %7s %5s %-16s %4s %6s %s\n",
+               "TIME", "PID", "UID", "COMM", "SIG", "TPID", "RESULT");
 }
 
 tracepoint:syscalls:sys_enter_kill
@@ -31,9 +31,9 @@ tracepoint:syscalls:sys_enter_kill
 tracepoint:syscalls:sys_exit_kill
 /@tpid[tid]/
 {
-       printf("%-15s %7d %-16s %4d %6d %6d\n",
+       printf("%-15s %7d %5d %-16s %4d %6d %6d\n",
                strftime("%H:%M:%S.%f", nsecs),
-               pid, comm, @tsig[tid], @tpid[tid], args.ret);
+               pid, uid, comm, @tsig[tid], @tpid[tid], args.ret);
        delete(@tpid[tid]);
        delete(@tsig[tid]);
 }
```

**Comment by DineshI-MS** [comment]
@viktormalik, thanks. This helps. Can I know what other properties can I pull here? I am actually looking for target process name as well. As the process is getting killed, I could not get that information. Also, another question. I tried to print to this table using **cat** command in the **_sys_enter_kill_** section. It worked. When I try to use multiple cat commands, it throws an error. Any suggestions on this. Using the cat command, I am trying to get process name and other info from path _**/proc/pid/**_

**Reply by viktormalik** [reply]
> I am actually looking for target process name as well. As the process is getting killed, I could not get that information. 

Unfortunately, not from bpftrace at the moment. The reason is that kernel allows this by using the `bpf_task_from_pid` kernel kfunc but bpftrace currently doesn't support kfuncs.

> Also, another question. I tried to print to this table using **cat** command in the **_sys_enter_kill_** section. It worked. When I try to use multiple cat commands, it throws an error. Any suggestions on this. Using the cat command, I am trying to get process name and other info from path _**/proc/pid/**_

Could you share a reproducer of the failure? It could be a bug.

**Reply by DineshI-MS** [reply]
@viktormalik, If I use 2 **cat** statements like below,
`
tracepoint:syscalls:sys_enter_kill
{   
	@tpid[tid] = pid;
	@tsig[tid] = args.sig;
	cat("/proc/%d/loginuid", pid) 
	cat("/proc/%d/comm", @tpid[tid])
}
`
I see this error
`
ERROR: syntax error, unexpected call, expecting }
    cat("/proc/%d/comm", @tpid[tid])
`

**Reply by viktormalik** [reply]
In this case, you're missing a semicolon (`;`) between the two commands.

**Reply by DineshI-MS** [reply]
Oh, my bad. It was a silly mistake. Thank you so much.

**Reply by DineshI-MS** [reply]
@viktormalik is there any way to read the file/ store output of cat command in a variable?

**Reply by viktormalik** [reply]
Unfortunately not. `cat` is an async command so the result cannot be stored in the BPF program, just printed.

**Reply by DineshI-MS** [reply]
@viktormalik can you suggest any way to read a file and store it in a variable? Or any other alternative to run a linux command and store the result in a file?

**Reply by viktormalik** [reply]
> @viktormalik can you suggest any way to read a file and store it in a variable?

No, not from a BPF program. If the contents of the file doesn't change during bpftrace execution, then you could dump its contents into a parameter:
```
# bpftrace -e 'BEGIN { $file = str($1); ... do something with $file ... }' $(cat file)
```
but keep in mind that the file contents will be stored as a string on the BPF stack which has a very limited capacity (512B), so this is only useful for small files.

> Or any other alternative to run a linux command and store the result in a file?

If you mean from bpftrace, you can run any linux command using the `system` builtin (need to use the unsafe mode, though):
```
# bpftrace -e 'BEGIN { system("echo hello > file"); }' --unsafe
```
